File: dynamic-programming/Knapsack.java
/**
* @author 	Luca Cortinovis	luca1cortinovis@gmail.com
* @version 	1.0
* @since	4/11/2022
*
* You can find me here:
* GitHub	Luca Cortinovis(NeptuneLuke)
*/


import java.util.Scanner;
import java.lang.Math;

public class Knapsack {

	
	public static void main(String args[]) {
		
		int values [] = {7,5,10,15,9,8,4};		// values (how many units of that element)
        int weights [] = {1,1,2,5,3,4,2}; 		// weights (how much each element weights)
        int W = 15;								// knapsack capacity (in units)
        
        int M[][]= knapsack(W,weights,values,values.length);
        
        printKnapsack(values.length,W,M,weights,values);
        
        System.out.println("Knapsack solution: " + M[values.length][W]);
	}
	
	
	/* This is the iterative algorithm for the Knapsack problem
	 * Time complexity: O(n*W)
	 * */
	public static int[][] knapsack(int W, int weights[], int values[], int n) {
		
		int M[][] = new int[n+1][W+1];
		
		// base case
		for(int i=0; i<=n; i++)
			M[i][0] = 0;
		
		// base case
		for(int w=0; w<=W; w++)
			M[0][w] = 0;
		
		// "recursive step"
		for(int i=1; i<=n; i++) {
			for(int j=1; j<=W; j++) {
				
				if(weights[i-1] > j)
					M[i][j] = M[i-1][j];
				else
					M[i][j] = Math.max(M[i-1][j], M[i-1][j-weights[i-1]] + values[i-1]);
			}
		}
		
		return M;
	}
	
	
	// Recursive algorithm to print the solution (with all subproblems)
	public static void printKnapsack (int i, int j, int M[][], int weights[], int values[]){
		
		if(i == 0 || j == 0)
			System.out.println();
		else {
			
			if(weights[i-1] > j)
				printKnapsack(i-1,j,M,weights,values);
			else {
				
				if(M[i-1][j-weights[i-1]] + values[i-1] >= M[i-1][j]) {
					printKnapsack(i-1,j-weights[i-1],M,weights,values);
					System.out.println("Object " + i + "   weight: " + weights[i-1] + "   value: " + values[i-1]);
				}
				else
					printKnapsack(i-1,j,M,weights,values);
			}
		}
		
	}
	
	
}

File: dynamic-programming/cpp/knapsack.cpp
/**
    * @author 	Luca Cortinovis	luca1cortinovis@gmail.com
    * @since	28/10/2023
    *
    * You can find me here:
    * GitHub	Luca Cortinovis (NeptuneLuke)
*/

#include <iostream>
#include <string>
#include <vector>


using
std::string,
std::vector,
std::cout,
std::cin;


void knapsack(int n, int W, vector<int> values, vector<int> weights, vector< vector<int> > &knapsack_table);
void print_knapsack(int i, int w, vector< vector<int> > knapsack_table, vector<int>values, vector<int> weights);


int main(void) {

    vector<int> values = {7, 5, 10, 15, 9, 8, 4};		// values (how many units of that element)
    vector<int> weights = {1, 1, 2, 5, 3, 4, 2}; 		// weights (how much each element weights)
    int W = 15;					                        // knapsack capacity

    int n = values.size();
    vector< vector<int> > knapsack_table (n + 1, vector<int> (W + 1) );

	print_knapsack(n, W, knapsack_table, values, weights);


    return 0;
}

void knapsack(int n, int W, vector<int> values, vector<int> weights, vector< vector<int> > &knapsack_table) {

	for (int i = 0; i <= n; ++i) {
		
		for (int j = 0; j <= W; ++j) {
			
			// base case i==0 or j==0
			if (i == 0 || j == 0) {
				knapsack_table[i][j] = 0;
			}

			// recursive step
			else {
				if (weights[i - 1] > j) {
					knapsack_table[i][j] = knapsack_table[i - 1][j];
				}
				else {
					knapsack_table[i][j] = std::max(knapsack_table[i - 1][j], knapsack_table[i - 1][j - weights[i - 1]] + values[i - 1]);
				}
			}
		}
	}

}

// i = object of index i
// w = remaining capacity of the knapsack
void print_knapsack(int i, int w, vector< vector<int> > knapsack_table, vector<int>values, vector<int> weights) {
	
	if (i == 0 || w == 0) {
		cout << "\n";
	}
	else {

		if (weights[i - 1] > w) {
			print_knapsack(i - 1, w, knapsack_table, values, weights);
		}
		else {

			if (knapsack_table[i - 1][w - weights[i - 1]] + values[i - 1] >= knapsack_table[i - 1][w]) {
				print_knapsack(i - 1, w - weights[i - 1], knapsack_table, weights, values);
				cout << "Object " << i << "   weight: " << weights[i - 1] << "   value: " << values[i - 1] << "\n";
			}
			else {
				print_knapsack(i - 1, w, knapsack_table, values, weights);
			}
		}
	}
}
